Index: l4/pkg/l4vfs/name_server/server/Makefile
===================================================================
--- l4/pkg/l4vfs/name_server/server/Makefile	(revision 473)
+++ l4/pkg/l4vfs/name_server/server/Makefile	(working copy)
@@ -12,6 +12,7 @@
 MODE          = l4env_minimal
 LIBS         += -ll4vfs_basic_name_server -ll4vfs_name_server-server
 
-DEFAULT_RELOC = 0x0151e000
+DEFAULT_RELOC = 0x01498000
+#DEFAULT_RELOC = 0x0151e000
 
 include $(L4DIR)/mk/prog.mk
Index: l4/pkg/l4vfs/simple_file_server/server/basic_io.c
===================================================================
--- l4/pkg/l4vfs/simple_file_server/server/basic_io.c	(revision 473)
+++ l4/pkg/l4vfs/simple_file_server/server/basic_io.c	(working copy)
@@ -192,6 +192,8 @@
         buf->st_mode  = S_IFDIR | S_IRWXU | S_IRWXG | S_IRWXO;
         buf->st_nlink = 1;
         buf->st_size  = 0;
+	//vs
+	buf->st_blksize = 512;
         ret = 0;
     }
     else
@@ -218,6 +220,8 @@
                         S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH;
         buf->st_nlink = 1;
         buf->st_size  = s_file->length;
+	//vs
+	buf->st_blksize = 512;
         ret = 0;
     }
 
Index: l4/pkg/l4vfs/simple_file_server/server/dirs.c
===================================================================
--- l4/pkg/l4vfs/simple_file_server/server/dirs.c	(revision 473)
+++ l4/pkg/l4vfs/simple_file_server/server/dirs.c	(working copy)
@@ -30,6 +30,9 @@
     // + 1 for the '\0'
     len = sizeof(temp->d_ino) + sizeof(temp->d_off) +
           sizeof(temp->d_reclen) + strlen(fname) + 1;
+#ifdef USE_UCLIBC
+    len += sizeof(temp->d_type);
+#endif
     // round up to align to word size
     len = (len + sizeof(int) - 1);
     len = (len / sizeof(int)) * sizeof(int);
@@ -44,7 +47,8 @@
                        // returnes slightly wrong number, at least for
                        // some FSs
     strncpy(entry->d_name, fname, L4VFS_MAX_NAME);
-    entry->d_name[L4VFS_MAX_NAME] = '\0';
+    //vs: entry->d_name[L4VFS_MAX_NAME] = '\0';
+    entry->d_name[strlen(fname)] = '\0';
 
     entry->d_reclen = dirent_size(fname);
 }
Index: l4/pkg/l4vfs/static_file_server/server/state.h
===================================================================
--- l4/pkg/l4vfs/static_file_server/server/state.h	(revision 473)
+++ l4/pkg/l4vfs/static_file_server/server/state.h	(working copy)
@@ -14,7 +14,8 @@
 
 #include <l4/sys/types.h>
 
-#define MAX_STATIC_FILES 32
+//#define MAX_STATIC_FILES 32
+#define MAX_STATIC_FILES 64
 #define STATIC_FILE_SERVER_VOLUME_ID 56
 /* Object-id 0 is reserved for the root dir
  * 1 .. MAX_STATIC_FILES + 1 for the files
Index: l4/pkg/l4vfs/term_server/server/term_con/Makefile
===================================================================
--- l4/pkg/l4vfs/term_server/server/term_con/Makefile	(revision 473)
+++ l4/pkg/l4vfs/term_server/server/term_con/Makefile	(working copy)
@@ -12,7 +12,8 @@
 
 TARGET         = term_con
 MODE           = l4env_minimal
-DEFAULT_RELOC  = 0x01400000
+#DEFAULT_RELOC  = 0x01400000
+DEFAULT_RELOC  = 0x01450000
 LIBS           = -u putchar -u puts -u printf                                \
                  -lcon -lvt100 -lcon_term -lconstream-server                 \
                  -ll4vfs_object_server-server -ll4vfs_name_server            \
