diff -urwN -x .git zip-os2-o/os2/os2.c zip-os2/os2/os2.c
--- zip-os2-o/os2/os2.c	2023-03-30 21:25:26.682243412 +0300
+++ zip-os2/os2/os2.c	2023-03-31 02:46:31.271356823 +0300
@@ -26,6 +26,10 @@
 #define PAD 0
 #define PATH_END '/'
 
+#ifdef __WATCOMC__
+#include <ctype.h>
+#endif
+
 
 #include "os2zip.h"
 
diff -urwN -x .git zip-os2-o/os2/os2zip.c zip-os2/os2/os2zip.c
--- zip-os2-o/os2/os2zip.c	2023-03-30 21:25:26.682243412 +0300
+++ zip-os2/os2/os2zip.c	2023-03-30 21:13:51.535579875 +0300
@@ -1052,18 +1052,20 @@
   struct stat s;
 
   strcpy( szName, z->name);
+#ifdef __KLIBC__
   // check if file is a symlink, query EAs from resolved file.
   if (linkput == 0 && lstat(z->name, &s) == 0)
   {
     if ((s.st_mode & S_IFLNK) == S_IFLNK)
       realpath( z->name, szName);
   }
+#endif
 
   /* store EA data in local header, and size only in central headers */
-  GetEAs(szName, &z->extra, &z->ext, &z->cextra, &z->cext);
+  GetEAs(szName, &z->extra, (size_t *)&z->ext, &z->cextra, (size_t *)&z->cext);
 
   /* store ACL data in local header, and size only in central headers */
-  GetACL(szName, &z->extra, &z->ext, &z->cextra, &z->cext);
+  GetACL(szName, &z->extra, (size_t *)&z->ext, &z->cextra, (size_t *)&z->cext);
 
 #ifdef USE_EF_UT_TIME
   /* store extended time stamps in both headers */
