diff --git a/dlls/ws2_32/socket16.c b/dlls/ws2_32/socket16.c
index 6b33263..8f551c5 100644
--- a/dlls/ws2_32/socket16.c
+++ b/dlls/ws2_32/socket16.c
@@ -294,8 +294,10 @@ INT16 WINAPI getsockopt16(SOCKET16 s, IN
     INT optlen32;
     INT *p = &optlen32;
     INT retVal;
-    if( optlen ) optlen32 = *optlen; else p = NULL;
-    retVal = WS_getsockopt( s, level, optname, optval, p );
+   /* since we're calling WS_getsockopt in 32bit, optlen is required to point to 4(bytes) */
+   /* we don't need a 32bit variable, but a variable that points to the size of a 32bit int.*/
+    if( optlen ) optlen32 = sizeof(INT); else p = NULL;
+    retVal = WS_getsockopt( s, (level & 0x00ffff), optname, optval, p );
     if( optlen ) *optlen = optlen32;
     return retVal;
 }
@@ -396,8 +398,11 @@ INT16 WINAPI sendto16(SOCKET16 s, char *
 INT16 WINAPI setsockopt16(SOCKET16 s, INT16 level, INT16 optname,
                           char *optval, INT16 optlen)
 {
+	INT optlen32 = sizeof(INT);
     if( !optval ) return SOCKET_ERROR;
-    return WS_setsockopt( s, level, optname, optval, optlen );
+   /* the win32 WS_setsockopt checks the optlen32, giving an optlen of 2 bytes
+    * causes an error */
+    return WS_setsockopt( s, (level & 0x00ffff), optname, optval, optlen32 );
 }
 
 /***********************************************************************
