diff -urN somfree/somipc/src/rhbsc.cpp somfree-new/somipc/src/rhbsc.cpp
--- somfree/somipc/src/rhbsc.cpp	(revision 95)
+++ somfree-new/somipc/src/rhbsc.cpp	(working copy)
@@ -252,7 +252,7 @@
 {
 	if (!p) p="unknown error";
 #ifdef _WIN32
-	fprintf(stderr,"SC: fatal: %s\n",p);
+        printf("SC: fatal: %s\n",p);
 	fflush(stdout);
 	fflush(stderr);
 #	ifdef _DEBUG
@@ -278,7 +278,7 @@
 		DebugStr("\p");
 	}
 #else
-	fprintf(stderr,"SC: fatal: %s\n",p);
+        printf("SC: fatal: %s\n",p);
 	fflush(stdout);
 	fflush(stderr);
 	exit(1);
@@ -473,8 +473,22 @@
 						}
 						else
 						{
-							fprintf(stderr,"%s: Unknown switch \'%s\'\n",appname,p);
-							return 1;
+                                                        if ((p[1]=='v'))
+                                                        {
+                                                                drop=1;
+                                                        }
+                                                        else
+                                                        {
+                                                                if ((p[1]=='c'))
+                                                                {
+                                                                        drop=1;
+                                                                }
+                                                                else
+                                                                {
+                                                                        fprintf(stderr,"%s: Unknown switch \'%s\'\n",appname,p);
+                                                                        return 1;
+                                                                }
+                                                        }
 						}
 					}
 				}
diff -urN somfree/somtklib/win32/somtklib.mak somfree-new/somtklib/win32/somtklib.mak
--- somfree/somtklib/win32/somtklib.mak	(revision 95)
+++ somfree-new/somtklib/win32/somtklib.mak	(working copy)
@@ -39,7 +39,9 @@
 		$(OUTDIR_DLL)\somem.lib \
 		$(OUTDIR_DLL)\somp.lib \
 		$(OUTDIR_DLL)\somnmf.lib \
-		$(OUTDIR_DLL)\soms.lib 
+                $(OUTDIR_DLL)\soms.lib \
+                $(OUTDIR_DLL)\somc.lib \
+                $(OUTDIR_DLL)\some.lib 
 
 $(SOMTKLIB): $(SOMLIBS)
 	if exist $@ del $@
