diff -urwN exehdr/src/exehdr.c newexehdr/src/exehdr.c
--- exehdr/src/exehdr.c	Thu Mar 14 10:14:03 2024
+++ newexehdr/src/exehdr.c	Thu Mar 14 10:10:24 2024
@@ -215,7 +215,7 @@
     if (optind == 1 || optind == 2)
     {
         // optind 1 should be input file
-        strncpy(exehdrinfo.infile, argv[optind], strlen(argv[optind]));
+        strncpy(exehdrinfo.infile, argv[optind], sizeof(exehdrinfo.infile)-1);
 
         if (access(exehdrinfo.infile, F_OK) != 0)
         {
@@ -233,7 +233,7 @@
 
         if (optind != argc)
             // provide output file
-            strncpy(exehdrinfo.outfile, argv[optind], strlen(argv[optind]));
+            strncpy(exehdrinfo.outfile, argv[optind], sizeof(exehdrinfo.outfile)-1);
         else
             // need to make an output file -- !!! template - do I need output file?
             sprintf(exehdrinfo.outfile, "%s%s", exehdrinfo.infname, ".txt");
@@ -588,9 +588,11 @@
         }
         va_end(ap);
     */
+	#ifdef __OS2__
     rc = DosGetMessage(0, (unsigned long)num_args, dataarea, MSG_BUFFER,
                        msgnum, "exh.msg", &msglen);
-
+	#else
+	#endif
     if (hdrdisp)
         prgheading();
 
diff -urwN exehdr/src/lxdisplay.c newexehdr/src/lxdisplay.c
--- exehdr/src/lxdisplay.c	Thu Mar 14 10:14:03 2024
+++ newexehdr/src/lxdisplay.c	Thu Mar 14 09:43:32 2024
@@ -630,7 +630,7 @@
                     printf("%s%s", spacefiller, tmpobjflag[flagcount]);
 
                     // if there is a second flag output it
-                    if (mod_count_obj_flags = 2)
+                    if (mod_count_obj_flags == 2)
                         printf(", %s", tmpobjflag[flagcount + 1]);
 
                     printf("\n");
