diff -urwN exehdr\src\exehdr.c newexehdr\src\exehdr.c
--- exehdr\src\exehdr.c	Thu Mar 14 09:41:57 2024
+++ newexehdr\src\exehdr.c	Thu Mar 14 09:34:45 2024
@@ -588,9 +588,11 @@
         }
         va_end(ap);
     */
+	#ifdef __OS2__
     rc = DosGetMessage(0, (unsigned long)num_args, dataarea, MSG_BUFFER,
                        msgnum, "exh.msg", &msglen);
-
+	#else
+	#endif
     if (hdrdisp)
         prgheading();
 
diff -urwN exehdr\src\lxdisplay.c newexehdr\src\lxdisplay.c
--- exehdr\src\lxdisplay.c	Thu Mar 14 09:41:57 2024
+++ newexehdr\src\lxdisplay.c	Thu Mar 14 09:17:39 2024
@@ -630,7 +630,7 @@
                     printf("%s%s", spacefiller, tmpobjflag[flagcount]);
 
                     // if there is a second flag output it
-                    if (mod_count_obj_flags = 2)
+                    if (mod_count_obj_flags == 2)
                         printf(", %s", tmpobjflag[flagcount + 1]);
 
                     printf("\n");
